django/db/backends/postgresql/operations.py

```diff
@@ -296,9 +296,17 @@ def distinct_sql(self, fields, params):
     if is_psycopg3:
 
         def last_executed_query(self, cursor, sql, params):
-            try:
-                return self.compose_sql(sql, params)
-            except errors.DataError:
+            conn_params = self.connection.get_connection_params()
+            cursor_type = conn_params["cursor_factory"].__name__
+            if cursor_type == "ServerBindingCursor":
```

Comment:
I'd check `server_side__binding` parameter instead
```suggestion
            server_side_binding = self.settings_dict["OPTIONS"].get("server_side_binding")
            if server_side_binding is True:
```